packages/medusa/src/strategies/batch-jobs/product/import.ts

```diff
@@ -0,0 +1,732 @@
+
+        writeStream.write(JSON.stringify(results[op]))
+      }
+    }
+
+    await Promise.all(uploadPromises)
+  }
+
+  /**
+   * Remove parsed ops JSON file.
+   *
+   * @param batchJobId - An id of the current batch job being processed.
+   * @param op - Type of import operation.
+   */
+  protected async downloadImportOpsFile(
+    batchJobId: string,
+    op: OperationType
+  ): Promise<TParsedRowData[]> {
+    let data = ""
+    const transactionManager = this.transactionManager_ ?? this.manager_
+
+    const readableStream = await this.fileService_
+      .withTransaction(transactionManager)
+      .getDownloadStream({
+        fileKey: `imports/products/import/ops/${batchJobId}-${op}`,
+        ext: "json",
+      })
+
+    return await new Promise((resolve) => {
+      readableStream.on("error", (a, b, c) => {
+        resolve([])
+      })
+      readableStream.on("data", (chunk) => {
+        data += chunk
+      })
+      readableStream.on("end", () => {
+        resolve(JSON.parse(data))
+      })
+    })
+  }
+
+  /**
+   * Delete parsed CSV ops files.
+   *
+   * @param batchJobId - An id of the current batch job being processed.
+   */
+  protected async deleteOpsFiles(batchJobId: string): Promise<void> {
+    const transactionManager = this.transactionManager_ ?? this.manager_
+
+    for (const op of Object.keys(OperationType)) {
+      try {
+        this.fileService_.withTransaction(transactionManager).delete({
+          fileKey: `imports/products/import/ops/-${batchJobId}-${op}`,
+        })
+      } catch (e) {
+        // noop
+      }
+    }
+  }
+
+  /**
+   * Update count of processed data in the batch job `result` column
+   * and cleanup temp JSON files.
+   *
+   * @param batchJobId - An id of the current batch job being processed.
+   */
+  private async finalize(batchJobId: string): Promise<void> {
+    const batchJob = await this.batchJobService_.retrieve(batchJobId)
+
+    delete this.processedCounter[batchJobId]
+
+    await this.batchJobService_.update(batchJobId, {
+      result: { progress: 1, advancement_count: batchJob.result.count },
+    })
+
+    const { fileKey } = batchJob.context as ImportJobContext
+
+    await this.fileService_.delete({ fileKey })
+
+    await this.deleteOpsFiles(batchJobId)
+  }
+
+  /**
+   * Store the progress in the batch job `result` column.
+   * Method is called after every update/create operation,
+   * but after every `BATCH_SIZE` processed rows info is written to the DB.
+   *
+   * @param batchJobId - An id of the current batch job being processed.
+   */
+  private async updateProgress(batchJobId: string): Promise<void> {
+    const newCount = (this.processedCounter[batchJobId] || 0) + 1
+    this.processedCounter[batchJobId] = newCount
+
+    if (newCount % BATCH_SIZE !== 0) {
+      return
+    }
+
+    await this.batchJobService_.update(batchJobId, {
+      result: {
+        advancement_count: newCount,
+        // progress: this.processedCounter / ,
```

Comment:
Suggestion: this seams to be not finished. Also, I am still not sure that the count/progress should depends on the class properties. Nothing should on my opinion. Same scenario as we discussed. If the server stopped (intentionally or not) we want to be able to restart where we stopped and therefore the job should hold the advancement etc in order to restart where it stopped otherwise the class will be reseted. Wdyt?

Reply:
Currently, progress counts are stored (and updated after a batch of 100 rows has been processed) in a map that happens to be a property of the class. Job recovery/restart is a separate feature that cannot be done by only using current count info in the case of the import strategy but rather needs to be additionally implemented as a feature. I think it is a good idea to ether somehow restart disrupted jobs or at least clean up ones that are stuck in the processing state, but since this functionality is not implemented in other batch job strategies I would suggest that we approach this as a separate issue, wdyt?

Also, regarding the progress count, what would be your suggestion?

Reply:
But i understand that for the import it would require a bit more data to recover and indeed it is another feature. We can let it that way for now and maybe keep a track of that somewhere ? Wdyt?

Reply:
yes let's keep using progress count in import context to display higher resolution progress of the job in the UI and I will create a ticker for "job recovery" implementation 👍

Reply:
👍